src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let newState = { ...state };
+  const stateHistory = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    switch (actions[i].type) {
+      case 'addProperties':
+        newState = { ...newState, ...actions[i].extraData };
+        stateHistory.push({ ...newState });
+        break;
+
+      case 'removeProperties':
+        for (let k = 0; k < actions[i].keysToRemove.length; k++) {
+          if (newState.hasOwnProperty(actions[i].keysToRemove[k])) {
+            delete newState[actions[i].keysToRemove[k]];
+          }
+        }
+
+        stateHistory.push({ ...newState });
+        break;
+
+      case 'clear':
+        newState = {};
+        stateHistory.push({ ...newState });
```

Comment:
Again, push the new state to the history array after the switch statement to avoid repetition.